File: go/core/message.go
package core

import "encoding/json"

// MessageStatus is the delivery status for one message.
type MessageStatus string

const (
	// StatusPending: accepted, not yet attempted.
	StatusPending MessageStatus = "pending"
	// StatusDelivering: attempt in flight.
	StatusDelivering MessageStatus = "delivering"
	// StatusDelivered: received a 2xx.
	StatusDelivered MessageStatus = "delivered"
	// StatusRetrying: transient failure, scheduled for a future attempt.
	StatusRetrying MessageStatus = "retrying"
	// StatusBlocked: endpoint circuit is open; parked, not counted as an attempt.
	StatusBlocked MessageStatus = "blocked"
	// StatusDead: moved to the DLQ (exhausted or permanent failure).
	StatusDead MessageStatus = "dead"
)

// Message is one (event x endpoint) delivery obligation. Fan-out turns one
// event into N messages, each with INDEPENDENT retry/circuit state (the key
// isolation invariant, G3). Mirrors TS Message (packages/core/src/types/message.ts).
type Message struct {
	MessageID  string
	Tenant     string
	EndpointID string
	EventID    string
	EventType  string
	Payload    json.RawMessage
	// AttemptNo is the 0-based attempt index into the retry schedule.
	AttemptNo int
	Status    MessageStatus
	// NextAttemptAt is the epoch-ms of the next scheduled attempt, set when
	// Status == StatusRetrying.
	NextAttemptAt *int64
	CreatedAt     int64
}

File: packages/core/src/types/message.ts
/** @fileoverview Message (event × endpoint delivery obligation) type. @module @anyhook/core */

/**
 * Delivery status for one message.
 * - `pending`    accepted, not yet attempted
 * - `delivering` attempt in flight
 * - `delivered`  received a 2xx
 * - `retrying`   transient failure, scheduled for a future attempt
 * - `blocked`    endpoint circuit is open; parked, not counted as an attempt
 * - `dead`       moved to the DLQ (exhausted or permanent failure)
 */
export type MessageStatus =
  | 'pending'
  | 'delivering'
  | 'delivered'
  | 'retrying'
  | 'blocked'
  | 'dead';

/**
 * One (event × endpoint) delivery obligation. Fan-out turns 1 event into N messages,
 * each with INDEPENDENT retry/circuit state (the key isolation invariant, §3).
 */
export interface Message {
  messageId: string;
  tenant: string;
  endpointId: string;
  eventId: string;
  eventType: string;
  payload: unknown;
  /** 0-based attempt index into the retry schedule. */
  attemptNo: number;
  status: MessageStatus;
  /** Epoch ms of the next scheduled attempt, when `status === 'retrying'`. */
  nextAttemptAt?: number;
  createdAt: number;
}
